ffcx/ir/integral.py

```diff
@@ -112,10 +113,22 @@ def compute_integral_ir(cell, integral_type, entitytype, integrands, argument_sh
                 if deps:
                     v['expression'] = v['expression']._ufl_expr_reconstruct_(*deps)
 
-            # Rebuild scalar target expressions and graph (this may be
-            # overkill and possible to optimize away if it turns out to be
-            # costly)
-            expression = S.nodes[S_targets[0]]['expression']
+            if value_rank == 0:
+                # Rebuild scalar target expressions and graph (this may be
+                # overkill and possible to optimize away if it turns out to be
+                # costly)
+                expression = S.nodes[S_targets[0]]['expression']
+            elif value_rank == 1 or value_rank == 2:
```

Comment:
What ufl function can we use for rank>2? 
rank 1: as_vector
rank 2: as_tensor
rank >2:?

Reply:
`as_tensor` is the most general function, because mathematically, so is "tensor". UFL uses special cases, if rank==1 then as_vector, if rank==2 then as_matrix. But I think you should be able all cases with as_tensor. It also suggests you should not use numpy's matrix below.